### ISSUE TITLE
Why inverse just part of VGG net(before relu3_1)?

### ISSUE THREAD

**Issue by TYLVY** [rtqichen/style-swap#16]
I just wonder if there is some way that the gradients can be backpropagated to the former layers, so that the net can be mush more powerful to extract features Phi(C) to swap with Phi(S)? I just want to try this, could you
please offer some suggestions on doing these? Thank you!

**Comment by rtqichen** [comment]
Do you mean the layers closer to the input? If the image is too large, then the swap operation takes a long time. But you can try it with smaller images and just set the `--layer` argument. eg. `--layer relu2_2` or something.

**Comment by TYLVY** [comment]
@rtqichen I don't mean that. You make a swap after relu3_1. The part net before 3_1 is fixed when training the inverse net. Then I wonder whether the fixed part net may limit the ability of the style transfer? Why not make  the part net before 3_1 adaptive to the specify task? In other words, here if I use C and S to respectively represent the output of content and style image after 3_1. Then after swapping, I get (C,S). When doing backpropagation, I wonder if d(C,S) can do some transformations to get d(C) (or the approximative transformations just to reduce the loss). I suppose that the dynamic part net before 3_1 can be more powerful for the task.

**Comment by rtqichen** [comment]
I see, so train it end-to-end rather than freezing the weights of the original network. This was done so it could generalize. Also, I didn't use the usual loss function from Gatys etal's work. In hind sight, maybe I should've just trained an inverse net to minimize that formulation. Though there are some follow up works that do train an end-to-end model that minimizes Gatys et al's formulation, though some tricks have to be employed so the network generalizes properly.

**Comment by TYLVY** [comment]
I see. I wonder that what are follow up works in  `there are some follow up works that do train an end-to-end model that minimizes Gatys et al's formulation` and some tricks if you don't mind...

**Comment by rtqichen** [comment]
I listed some in #15 but haven't looked into this since then.
